src/LinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -98,12 +99,80 @@ void UprobesUnwindingVisitor::visit(CallchainSamplePerfEvent* event) {
+  // function is never the outermost frame. For example, for the main thread the outermost function
+  // is _start, followed by __libc_start_main. For other threads, the outermost function is clone.
+  if (event->GetCallchainSize() <= 2) {
+    if (unwind_error_counter_ != nullptr) {
+      ++(*unwind_error_counter_);
+    }
+    return;
+  }
+
+  // The leaf function is not guaranteed to have the frame pointer for all our targets. Though, we
+  // assume that $rbp remains untouched by the leaf functions, such that we can rely on
+  // perf_event_open to give us "almost" correct callstacks (the caller of the leaf function will be
+  // missing). We can sanity check this assumption by checking if the callstack only contains
+  // executable code.
+  // TODO(b/187690455): As soon as we actually have frame pointers in all non-leaf functions, we can
+  //  "discard" the sample. Till that point, this check will always fail, as libc does not have
+  //  frame-pointers, the caller of __libc_start_main will be invalid.
+  //  Note, that at this point in time, we don't want to actually throw the sample away, but rather
+  //  report it as "broken".
+  for (uint64_t frame_index = 1; frame_index < event->GetCallchainSize(); ++frame_index) {
+    unwindstack::MapInfo* map_info = current_maps_->Find(event->GetCallchain()[frame_index]);
+    if (map_info == nullptr || (map_info->flags & PROT_EXEC) == 0) {
+      ERROR("$rbp was modified.");
+      break;
+    }
+  }
+
+  std::optional<uint64_t> leaf_function_caller_address;
+
+  // Unwind the first two frames with libunwindstack. If the outer frame (i.e. the caller of the
+  // function at $pc) is different to the one reported by perf_event_open, it was a leaf function,
+  // and we need to add this frame to perf_event_open's callstack.
+  const LibunwindstackCallstack& libunwindstack_result =
+      unwinder_.Unwind(event->GetPid(), current_maps_.get(), event->GetRegisters(),
+                       event->GetStackData(), event->GetStackSize(), 2);
+  const std::vector<unwindstack::FrameData>& libunwindstack_callstack =
+      libunwindstack_result.frames;
+
+  // We expect exactly two frames (as we've set `max_frames` to 2), otherwise it is an unwinding
+  // error.
+  // Note that this doesn't exclude samples inside the main function of any thread as the main
+  // function is never the outermost frame. For example, for the main thread the outermost function
+  // is _start, followed by __libc_start_main. For other threads, the outermost function is clone.
+  if (libunwindstack_callstack.size() != 2) {
+    if (unwind_error_counter_ != nullptr) {
+      ++(*unwind_error_counter_);
+    }
+    return;
+  }
+
+  // Note, both, frame-pointer-based unwinding and libunwindstack use the return address of a
+  // function call as the caller's address (which is the instruction after the call instruction), as
+  // they don't know the size of the call instruction. To fix this, libunwindstack decreases the
```

Comment:
"this" here refers to the fact that the caller used is the instruction after the call instruction, but it's not really clear, as that's not the core of the previous sentence, rather it's in parenthesis. I think this should be slightly reformulated.